File: script/make/protocol/lua/MailProtocol.lua
MailProtocol = {}

function MailProtocol.encode(offset, protocol, data)
    if protocol == 11401 then
        local offset = offset
        local table = {}
        return table
    elseif protocol == 11402 then
        local offset = offset
        local table = {}
        -- 邮件ID
        table[offset] = string.pack(">I8", data["mailId"])
        offset = offset + 1
        return table
    elseif protocol == 11403 then
        local offset = offset
        local table = {}
        -- 邮件ID
        table[offset] = string.pack(">I8", data["mailId"])
        offset = offset + 1
        return table
    elseif protocol == 11404 then
        local offset = offset
        local table = {}
        -- 邮件ID
        table[offset] = string.pack(">I8", data["mailId"])
        offset = offset + 1
        return table
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

function MailProtocol.decode(offset, protocol, data)
    if protocol == 11401 then
        local offset = offset
        -- 邮件列表
        local list = {}
        local listLength = string.unpack(">I2", data, offset)
        offset = offset + 2
        for listIndex = 1, listLength do
            -- 邮件ID
            local mailId = string.unpack(">I8", data, offset)
            offset = offset + 8
            -- 接收时间
            local receiveTime = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 有效时间
            local expireTime = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 读取时间
            local readTime = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 领取附件时间
            local receiveAttachmentTime = string.unpack(">I4", data, offset)
            offset = offset + 4
            -- 标题
            local title = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(title)
            -- 内容
            local content = string.unpack(">s2", data, offset)
            offset = offset + 2 + string.len(content)
            -- 附件列表
            local attachment = {}
            local attachmentLength = string.unpack(">I2", data, offset)
            offset = offset + 2
            for attachmentIndex = 1, attachmentLength do
                -- 物品ID
                local itemId = string.unpack(">I4", data, offset)
                offset = offset + 4
                -- 数量
                local number = string.unpack(">I2", data, offset)
                offset = offset + 2
                attachment[attachmentIndex] = {itemId = itemId, number = number}
            end
            list[listIndex] = {mailId = mailId, receiveTime = receiveTime, expireTime = expireTime, readTime = readTime, receiveAttachmentTime = receiveAttachmentTime, title = title, content = content, attachment = attachment}
        end
        return {list = list}
    elseif protocol == 11402 then
        local offset = offset
        -- 结果
        local result = string.unpack(">s2", data, offset)
        offset = offset + 2 + string.len(result)
        return {result = result}
    elseif protocol == 11403 then
        local offset = offset
        -- 结果
        local result = string.unpack(">s2", data, offset)
        offset = offset + 2 + string.len(result)
        return {result = result}
    elseif protocol == 11404 then
        local offset = offset
        -- 结果
        local result = string.unpack(">s2", data, offset)
        offset = offset + 2 + string.len(result)
        return {result = result}
    else
        error(string.format('unknown protocol define: %d', protocol))
    end
end

File: script/make/protocol/cs/MailProtocol.cs
public static class MailProtocol
{
    public static void Encode(System.Text.Encoding encoding, System.IO.BinaryWriter writer, System.UInt16 protocol, System.Object data) 
    {
        switch (protocol) 
        {
            case 11401:
            {
                return;
            }
            case 11402:
            {
                // 邮件ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int64)(System.UInt64)data));
                return;
            }
            case 11403:
            {
                // 邮件ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int64)(System.UInt64)data));
                return;
            }
            case 11404:
            {
                // 邮件ID
                writer.Write(System.Net.IPAddress.HostToNetworkOrder((System.Int64)(System.UInt64)data));
                return;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }

    public static System.Object Decode(System.Text.Encoding encoding, System.IO.BinaryReader reader, System.UInt16 protocol) 
    {
        switch (protocol) 
        {
            case 11401:
            {
                // 
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = new System.Collections.Generic.List<System.Object>(dataLength);
                while (dataLength-- > 0)
                {
                    // 
                    // 邮件ID
                    var mailId = (System.UInt64)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt64());
                    // 接收时间
                    var receiveTime = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 有效时间
                    var expireTime = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 读取时间
                    var readTime = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 领取附件时间
                    var receiveAttachmentTime = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                    // 标题
                    var titleLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var title = encoding.GetString(reader.ReadBytes(titleLength));
                    // 内容
                    var contentLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var content = encoding.GetString(reader.ReadBytes(contentLength));
                    // 附件列表
                    var attachmentLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                    var attachment = new System.Collections.Generic.List<System.Object>(attachmentLength);
                    while (attachmentLength-- > 0)
                    {
                        // 
                        // 物品ID
                        var attachmentItemId = (System.UInt32)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt32());
                        // 数量
                        var attachmentNumber = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                        // object
                        var attachmentItem = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"item_id", attachmentItemId}, {"number", attachmentNumber}};
                        // add
                        attachment.Add(attachmentItem);
                    }
                    // object
                    var mail = new System.Collections.Generic.Dictionary<System.String, System.Object>() {{"mail_id", mailId}, {"receive_time", receiveTime}, {"expire_time", expireTime}, {"read_time", readTime}, {"receive_attachment_time", receiveAttachmentTime}, {"title", title}, {"content", content}, {"attachment", attachment}};
                    // add
                    data.Add(mail);
                }
                return data;
            }
            case 11402:
            {
                // 结果
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = encoding.GetString(reader.ReadBytes(dataLength));
                return data;
            }
            case 11403:
            {
                // 结果
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = encoding.GetString(reader.ReadBytes(dataLength));
                return data;
            }
            case 11404:
            {
                // 结果
                var dataLength = (System.UInt16)System.Net.IPAddress.NetworkToHostOrder(reader.ReadInt16());
                var data = encoding.GetString(reader.ReadBytes(dataLength));
                return data;
            }
            default:throw new System.ArgumentException(System.String.Format("unknown protocol define: {0}", protocol));
        }
    }
}